### PR TITLE
Kill RAnal.sdb_fcns

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**DO NOT SQUASH!!!!!!!!!!!!!!!!!!!!!!!!!!!!**

Filling this template is power, France is bacon.

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

### Detailed description

I'll go through the previously remaining uses of `sdb_fcns` and how they have been killed.

#### Variables

... were saved in this sdb until #16547 which was merged months ago. Apparently the only remaining use of vars there was the `exists_var` callback for RPrint, which I rewrote for the correct variable storage.

#### Variable XRefs

`axt` queried accesses to variables from `sdb_fmt`, but there was no place that actually wrote these accesses there so it was basically a nop. The correct way to query accesses to vars is by using `afv=`.

#### Type propagation

This one is a bit more complicated. This is roughly how the relevant type matching part worked before:

```
foreach f in all functions {
    foreach call to callee in f {
        if callee is a user-defined function {
            sdb_fcns["arg i of callee has type"] = "type coming from f's vars" foreach i in len(callee args)
        } else { ... }
    }
    foreach i where "arg i of f has type" in sdb_fcns {
        f.arg[i].type = sdb_fcns["arg i of f has type"]
    }
}
```

So whenever types should be propagated from a caller into a callee's args, they weren't directly set, but rather temporarily saved by abusing `sdb_fmt` and only later actually applied to the callee. Do not confuse `f` and `callee` above! In the bottom loop, the types for `f` are applied, which are generally not the ones saved in the same loop iteration of the outermost loop.
This had the negative side-effect that some callee's just wouldn't get their args propagated, depending on their position in the list of functions.

I changed it to the more intuitive solution, just applying the types directly:
```
foreach f in all functions {
    foreach call to callee in f {
        if callee is a user-defined function {
            callee.arg[i].type = "type coming from f's vars" foreach i in len(callee args)
        } else { ... }
    }
}
```

This gets rid of the abuse of the global sdb as local temporary variables and fixes the order-dependence mentioned above.
The fixed order-dependence does slightly change the results of a few tests. I checked them in detail and some now have better results, some a bit worse with the worse results being the result of bugs in the backtracking of types and ignoring of pointer derefs and refs, but that part is not touched by this pr so it is overall more correct now.

**tl;dr**: The sdb was abused for local storage, this was fixed by fixing a bug, which affects some tests positively and some slightly negatively, but overall it's better now and definitely more correct.

### Test plan

run the tests

### CODE REVIEW

File: libr/core/anal_tp.c [hunk 1]
```diff
@@ -412,6 +449,7 @@ static int bb_cmpaddr(const void *_a, const void *_b) {
 }
 
 R_API void r_core_anal_type_match(RCore *core, RAnalFunction *fcn) {
+	eprintf ("r_core_anal_type_match (%s @ 0x%"PFMT64x")\n", fcn->name, fcn->addr);
```
Comment (trufae) [inline libr/core/anal_tp.c:452]: r_return_if_fail

File: libr/core/anal_tp.c [hunk 2]
```diff
@@ -235,12 +235,52 @@ static RList *parse_format(RCore *core, char *fmt) {
...
+ * \param addr addr of the call instruction
+ * \param baddr addr of the caller function
+ * \param cc cc of the callee
+ * \param prev_idx index in the esil trace
+ * \param userfnc whether the callee is a user function (affects propagation direction)
+ * \param caddr addr of the callee
+ */
 static void type_match(RCore *core, char *fcn_name, ut64 addr, ut64 baddr, const char* cc,
 		int prev_idx, bool userfnc, ut64 caddr) {
+	eprintf ("  type_match(%s @ 0x%"PFMT64x", baddr: 0x%"PFMT64x", userfnc: %d, 0x%"PFMT64x")\n", fcn_name, addr, baddr, userfnc, caddr);
```
Comment (trufae) [inline libr/core/anal_tp.c:283]: remove debug message

File: libr/core/anal_tp.c (hunk 1 above)
Comment (thestr4ng3r) [reply libr/core/anal_tp.c:452]: This is all debug stuff, I am still testing.

File: test/db/cmd/types [hunk 3]
```diff
@@ -1260,7 +1260,7 @@ var char * var_28h @ rbp-0x28
 var char * var_20h @ rbp-0x20
 var int64_t var_18h @ rbp-0x18
 var int64_t var_8h @ rbp-0x8
-arg char ** arg2 @ rsi
+arg char * arg2 @ rsi
```
Comment (trufae) [inline test/db/cmd/types:1263]: is this change correct?

**Review by trufae — APPROVED** [review]
LGTM, just wondering about the char ** -> char * change if its a fix or a regression

**Review by XVilka — APPROVED** [review]

**Comment by trufae** [comment]
Newshell red

**Comment by trufae** [comment]
Redshell new

File: test/db/cmd/types (hunk 3 above)
Comment (thestr4ng3r) [reply test/db/cmd/types:1263]: It is a slightly worse result for how the code looks here but with respect to how the algorithm works it is correct, so yes it's a fix.
